infra/Pos/Slotting/Types.hs

```diff
@@ -22,15 +38,123 @@ data EpochSlottingData = EpochSlottingData
+
+    nextEpochSlottingData :: Map EpochIndex NextEpochSlottingData
+    nextEpochSlottingData = M.singleton 1 esd
+
+-- | Get all epoch index.
+getAllEpochIndices :: SlottingData -> [EpochIndex]
+getAllEpochIndices = M.keys . getSlottingDataMap
+
+-- | Get the next epoch index.
+getNextEpochIndex :: SlottingData -> EpochIndex
+getNextEpochIndex = fst . M.findMax . getSlottingDataMap
+
+-- | Get the next epoch slotting data.
+getNextEpochSlottingData :: SlottingData -> EpochSlottingData
+getNextEpochSlottingData = snd . M.findMax . getSlottingDataMap
+
+-- | Get the current epoch index. Next epoch - 1.
+getCurrentEpochIndex :: SlottingData -> EpochIndex
+getCurrentEpochIndex = decreaseEpochIndex . getNextEpochIndex
+  where
+    -- Left for readability.
+    decreaseEpochIndex :: EpochIndex -> EpochIndex
+    decreaseEpochIndex = pred
+
+-- | Get the current epoch slotting data. Next epoch - 1.
+getCurrentEpochSlottingData :: SlottingData -> EpochSlottingData
+getCurrentEpochSlottingData sdp@(getSlottingDataMap -> sd) = sd M.! currentEpochIndex
+  where
+    currentEpochIndex = getCurrentEpochIndex sdp
+
+-- | Lookup the slotting data for an arbitrary `EpochIndex`.
+lookupEpochSlottingData :: EpochIndex -> SlottingData -> Maybe EpochSlottingData
+lookupEpochSlottingData epochIndex slottingData = M.lookup epochIndex slottingData'
+  where
+    slottingData' :: Map EpochIndex EpochSlottingData
+    slottingData' = getSlottingDataMap slottingData
+
+-- | Insert `EpochSlottingData`.
+addEpochSlottingData :: EpochIndex -> EpochSlottingData -> SlottingData -> SlottingData
+addEpochSlottingData epochIndex epochSlottingData slottingData =
+    SlottingData $ M.insert epochIndex epochSlottingData slottingData'
+  where
+    slottingData' :: Map EpochIndex EpochSlottingData
+    slottingData' = getSlottingDataMap slottingData
+
+-- | Compute when the slot started. We give it @SlotId@, the @SlotId@
+-- @EpochSlottingData@ and find when did that @SlotId@ occur.
+computeSlotStart :: Timestamp -> SlotId -> EpochSlottingData -> Timestamp
+computeSlotStart systemStart slotId esd =
+      slotTimestamp systemStart siLocalSlotIndex esd
+    where
+      siLocalSlotIndex :: LocalSlotIndex
+      siLocalSlotIndex = siSlot slotId
+
+      slotTimestamp
+          :: Timestamp
+          -> LocalSlotIndex
+          -> EpochSlottingData
+          -> Timestamp
+      slotTimestamp systemStart' localSlotIndex' epochSlottingData =
+          epochStartTime + currentSlotTimestamp
+        where
+          intSlotIndex :: Word16
+          intSlotIndex = getSlotIndex localSlotIndex'
+
+          -- We get the epoch start time by adding the epoch slotting data start diff
+          -- which is:
+          --   currentEpochStart - systemStart + systemStart = currentEpochStart
+          --
+          -- Seems kind of dubious.
+          epochStartTime :: Timestamp
+          epochStartTime = addTimeDiffToTimestamp epochStartTimeDiff systemStart'
+            where
```

Comment:
Nested `where`s make code less understandable. I think it's better to go with `let` if you don't want to move declaration out of function scope instead of nesting `where`.
In this particular case as `slotTimestamp` has the same arguments as `computeSlotStart` you can move `intSlotIndex` outside to the top level `where` for example.